File: DesignPatterns/01_GOF23/DP_examples/C/03_Behavioral/Strategy/context.c
#include "context.h"

// 初始化上下文
void Context_init(Context* self, IStrategy* strategy) {
    self->strategy = strategy;
}

// 设置策略
void Context_setStrategy(Context* self, IStrategy* strategy) {
    self->strategy = strategy;
}

// 执行排序
void Context_sort(Context* self, int* array, int length) {
    if (self->strategy && self->strategy->sort) {
        self->strategy->sort(self->strategy, array, length);
    }
}

File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/Strategy/context.rs
// 上下文类
use crate::strategy::IStrategy;

pub struct Context {
    strategy: Box<dyn IStrategy>,
}

impl Context {
    pub fn new(strategy: Box<dyn IStrategy>) -> Self {
        Self {
            strategy,
        }
    }
    
    pub fn set_strategy(&mut self, strategy: Box<dyn IStrategy>) {
        self.strategy = strategy;
    }
    
    pub fn sort(&self, array: &mut [i32]) {
        self.strategy.sort(array);
    }
}
